apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerCommandTextBuilder.cs

```diff
@@ -255,13 +255,117 @@ public static SpannerCommandTextBuilder FromCommandText(string commandText)
+                                $"Spanner {commandName} commands may be specified as '{commandName} <table>' with " +
+                                "parameters added to customize the command with filtering or updated values. " +
+                                "In that case, the commands are sent to Spanner as mutations which don't support comments" +
+                                "or hints, or any other SQL syntax feature. If you need to use comments or hints, " +
+                                "please use DML commands. You can read more about this topic in " +
+                                "https://cloud.google.com/spanner/docs/dml-versus-mutations.");
+                        }
                         targetTable = ValidateTableName(commandSections[1], nameof(commandText));
                         // This just normalizes the command text.
                         commandText = $"{commandName} {targetTable}";
                     }
                     break;
             }
-            return new SpannerCommandTextBuilder(commandText, commandType, targetTable, extraStatements: null);
+            return new SpannerCommandTextBuilder(commandText.Trim(), commandType, targetTable, extraStatements: null);
+        }
+
+        private static string RemoveLeadingCommentsAndHints(string commandText, out bool removedCommentOrHint)
+        {
+            try
+            {
+                return commandText.Substring(IndexOfCommandStart(commandText, out removedCommentOrHint));
+            }
+            catch(Exception ex) when (ex is ArgumentOutOfRangeException || ex is IndexOutOfRangeException)
+            {
+                throw new ArgumentException($"'{commandText}' is not a recognized Spanner command.", nameof(commandText), ex);
+            }
+        }
+
+        private static int IndexOfCommandStart(string command, out bool removedCommentOrHint)
+        {
+            int currentIndex = 0;
+            int loops = 0;
+            do
+            {
+                loops++;
+                SkipWhiteSpaces();
+            }
+            while (SkipLineComment() || SkipBlockComment() || SkipHint());
+
+            removedCommentOrHint = loops > 1;
+            return currentIndex;
+
+            void SkipWhiteSpaces()
+            {
+                while (char.IsWhiteSpace(command[currentIndex]))
+                {
+                    currentIndex++;
+                }
+            }
+
+            bool SkipLineComment()
+            {
+                if (command.IndexOf('#', currentIndex, 1) >= 0 ||
+                    command.IndexOf("--", currentIndex, 2) >= 0)
+                {
+                    // TODO: If this can be escaped, let's check that it's not
+                    // or continue searching.
+                    currentIndex = command.IndexOf('\n', currentIndex);
+                    if (currentIndex >= 0) // If we didn't find it, we don't fail here.
```

Comment:
I don't understand the comment here. I'd expect something like "If there are more lines, move to the start of the next line and continue".

Reply:
This is just: we didn't find an end of line, so the whole thing is a comment, as we are skipping a line comment. currentIndex will be -1 and the next time we try to use it it will throw ArgumentOutOfRange or IndexOutOfRange which are caught by the caller to throw ArgumentException("unrecognized command").

I'm fine with throwing here something more explicit though.